lib/storage/src/content_manager/consensus/consensus_wal.rs

```diff
@@ -1,203 +1,288 @@
+            entries.push(entry);
+
+            if let Some(max_size_bytes) = max_size_bytes {
+                if size_bytes >= max_size_bytes {
+                    break;
+                }
+            }
+        }
 
         Ok(entries)
     }
 
     pub fn first_entry(&self) -> Result<Option<RaftEntry>, StorageError> {
-        let first_index = self.0.first_index();
-        let entry = self
-            .0
-            .entry(first_index)
-            .map(|entry| <RaftEntry as prost_for_raft::Message>::decode(entry.as_ref()));
-        Ok(entry.transpose()?)
+        self.entry_by_wal_index_impl(self.wal.first_index())
     }
 
     pub fn last_entry(&self) -> Result<Option<RaftEntry>, StorageError> {
-        let last_index = self.0.last_index();
-        let entry = self
-            .0
-            .entry(last_index)
-            .map(|entry| <RaftEntry as prost_for_raft::Message>::decode(entry.as_ref()));
-        Ok(entry.transpose()?)
+        self.entry_by_wal_index_impl(self.wal.last_index())
     }
 
-    /// Difference between raft index and WAL record number.
-    /// Difference might be different because of consensus snapshot.
-    pub fn index_offset(&self) -> Result<Option<u64>, StorageError> {
-        let last_known_index = self.0.first_index();
-        let first_entry = self.first_entry()?;
-        let offset = first_entry.map(|entry| entry.index - last_known_index);
-        Ok(offset)
-    }
+    pub fn append_entries(&mut self, new_entries: Vec<RaftEntry>) -> Result<(), StorageError> {
+        if new_entries.is_empty() {
+            return Ok(());
+        }
 
-    pub fn append_entries(&mut self, entries: Vec<RaftEntry>) -> Result<(), StorageError> {
-        for entry in entries {
-            let operation_opt = ConsensusOperations::try_from(&entry).ok();
+        // Calculate WAL to Raft index offset
+        let mut current_index_offset = self.index_offset_impl()?;
+
+        // Use single buffer to encode all new entries to reduce allocations
+        let mut buf = Vec::new();
+
+        for new_entry in new_entries {
+            // If WAL is not empty, check that new entry index is within WAL bounds
+            if let Some(offset) = current_index_offset {
+                // Check that new entry index was not already compacted (it's not less than first WAL index)
+                let Some(new_entry_wal_index) = offset.try_raft_to_wal(new_entry.index) else {
+                    return Err(StorageError::service_error(format!(
+                        "Can't append entry with Raft index {}, \
+                         because WAL is already compacted at Raft index {}",
+                        new_entry.index, offset.raft_index,
+                    )));
+                };
 
-            let index = entry.index;
-            let current_index = self.0.last_index();
-            let index_offset = self.index_offset()?;
+                let next_wal_index = self.wal.last_index() + 1;
```

Comment:
Note that if we have either 0 or 1 items in our WAL, `last_index()` produces `0` in both cases.

In other words, it returns a nonsensical value if the WAL is still empty. I'm not sure if that's desirable.

Maybe we want to return an option there so we have to handle this case explicitly.

Reply:
This whole branch is only triggered when `current_index_offset` is `Some`, which means there is at least one entry in the WAL, which means `Wal::last_index` should be valid.

Reply:
This whole branch is only triggered when `current_index_offset` is `Some`, which means there is at least one entry in the WAL, which means `Wal::last_index` should be valid.

Reply:
Makes sense, thanks!